packages/gatsby/src/utils/start-server.ts

```diff
@@ -487,9 +488,110 @@ module.exports = {
 
   // Render an HTML page and serve it.
   if (process.env.GATSBY_EXPERIMENTAL_DEV_SSR) {
-    // Setup HTML route.
-    const { route } = require(`./dev-ssr/develop-html-route`)
-    route({ app, program, store })
+    app.get(`*`, async (req, res, next) => {
+      telemetry.trackFeatureIsUsed(`GATSBY_EXPERIMENTAL_DEV_SSR`)
+
+      const pathObj = findPageByPath(store.getState(), decodeURI(req.path))
+
+      if (!pathObj) {
+        return next()
+      }
+
+      await appendPreloadHeaders(pathObj.path, res)
+
+      const htmlActivity = report.phantomActivity(`building HTML for path`, {})
+      htmlActivity.start()
+
+      try {
+        const renderResponse = await renderDevHTML({
+          path: pathObj.path,
+          page: pathObj,
+          skipSsr: req.query[`skip-ssr`] || false,
+          store,
+          htmlComponentRendererPath: `${program.directory}/public/render-page.js`,
+          directory: program.directory,
+        })
+        res.status(200).send(renderResponse)
+      } catch (error) {
+        // The page errored but couldn't read the page component.
+        // This is a race condition when a page is deleted but its requested
+        // immediately after before anything can recompile.
+        if (error === `404 page`) {
+          return next()
+        }
+
+        // renderDevHTML throws an error with these information
+        const lineNumber = error?.line
+        const columnNumber = error?.column
+        const filePath = error?.filename
+        const sourceContent = error?.code
+
+        report.error({
+          id: `11614`,
+          context: {
+            path: pathObj.path,
+            filePath: filePath,
+            line: lineNumber,
+            column: columnNumber,
+          },
+        })
+
+        const emptyResponse = {
+          isDevSSRError: true,
+          codeFrame: `No codeFrame could be generated`,
+          sourcePosition: null,
+          sourceContent: null,
+        }
+
+        if (!sourceContent || !lineNumber) {
+          res.json(emptyResponse)
+        }
+
+        const codeFrame = codeFrameColumns(
+          sourceContent,
+          {
+            start: {
+              line: lineNumber,
+              column: columnNumber ?? 0,
+            },
+          },
+          {
+            highlightCode: true,
+          }
+        )
+
+        const message = {
+          codeFrame,
+          source: filePath,
+          line: lineNumber,
+          column: columnNumber ?? 0,
+        }
+
+        // Generate a shell for client-only content -- for the error overlay
+        const clientOnlyShell = await renderDevHTML({
+          path: pathObj.path,
+          page: pathObj,
+          skipSsr: true,
+          store,
+          htmlComponentRendererPath: `${program.directory}/public/render-page.js`,
+          directory: program.directory,
+        })
```

Comment:
What if we add a new object to renderDevHTML that contain the error message so `renderDevHTML` can include it in react's rendering piece.

Reply:
Certainly less hackier than what we have right now 👍 

So we'd pass `message` to `renderDevHTML` and it would take care of the pushing to `window._gatsbyEvents` or would be pass an object like:

```js
error: {
  action: 'SHOW_DEV_SSR_ERROR',
  payload: message,
}
```

Reply:
I would personally do `error: message`. It doesn't need to know implementation details of SHOW_DEV_SSR_ERROR

Reply:
Done!